CMakeLists.txt

```
@@ -136,6 +138,8 @@ else()
 	endif()
 endif()
 
+target_compile_features(libninja PUBLIC cxx_std_11)
```

Isn't it better to use the most recent C++ standard when it is not specified manually? 

If someone needs an older standard like c++11, they can pass `-D CMAKE_CXX_STANDARD=11`.

https://github.com/cpp-best-practices/project_options/blob/2111be3ee3c1e74ad7d38e25a51f33bff06f3dc1/src/Standards.cmake#L1

---

Ideally compiling Ninja should work out of the box even on old systems like RHEL 7.

You're right that this might disable some optimizations on newer systems. Too bad CMake doesn't have a way to use the latest standard available?

---

> Too bad CMake doesn't have a way to use the latest standard available?

It does if you check the link I gave. The function is part this library:
https://github.com/cpp-best-practices/project_options/

---

I was thinking more of a built-in way. In that function the standards are still hard-coded, so it only tells the truth as of right now. What happens when C++23 is released? The CMake code needs to be updated.

Sorry, but I don't like extending the CMakeLists.txt too much. `target_compile_features(libninja PUBLIC cxx_std_11)` is one line which is easy enough to understand. It would be better to tell CMake that we need *at least* C++11, but as there's no easy built-in way to do that we're stuck with that.

---

> I was thinking more of a built-in way


The defaults for CMake and compilers are not the best at all. That's the whole reason we created `project_options`. 


> In that function the standards are still hard-coded, so it only tells the truth as of right now. What happens when C++23 is released? 

This is not hard coding. It is providing a range of versions that the authors of the library support. When C++23 is released, first you should try building with it, and once you are sure it works, you can add that to the version range.

> is one line which is easy enough to understand

It is your choice to stick to these sub-optimal (sometimes bad defaults). I don't think a `if` statement is considered complex in the C++ world. 

Worst case scenario, you should at least check if the standard is not set manually by the user:
```suggestion
if("${CMAKE_CXX_STANDARD) STREQUAL "")
	set(CMAKE_CXX_STANDARD 11)
endif()
```

---

I just found [CMAKE_CXX_STANDARD_DEFAULT](https://gitlab.kitware.com/cmake/cmake/-/issues/21115). Maybe we can do the following:

```cmake
if(CMAKE_CXX_STANDARD_DEFAULT EQUAL 98)
  target_compile_features(libninja PUBLIC cxx_std_11)
endif()
```

Haven't tested that, but the idea would be not to mess with the compilers default unless we have to.